src/core/mount.c

```diff
@@ -823,6 +826,10 @@ static void mount_enter_dead(Mount *m, MountResult f) {
         unit_unref_uid_gid(UNIT(m), true);
 
         dynamic_creds_destroy(&m->dynamic_creds);
+
+        /* Any information and dependencies based on /proc/self/mountinfo are now stale */
+        assert_se(update_parameters_proc_self_mount_info(m, NULL, NULL, NULL) >= 0);
+        unit_remove_dependencies(UNIT(m), UNIT_DEPENDENCY_MOUNTINFO_IMPLICIT);
```

Comment:
i don't grok this part. why is this necessary? we read /proc/self/mountinfo anyway very soon after, which should drop the data too, no? what's the reason for pulling that into this call, ahead of time?

Reply:
I assume you are talking about the call to `update_parameters_...`. I tried to explain this in the commit message, but I know I wasn't very clear. The calls to `unit_remove_dependnecies` removes the references to the loopback device from the mount unit. This fixes the original bug reported in #11342. But an opposite issue starts occuring: after a sequence of start, stop, start of the mount unit, the mount doesn't have the automatic dependencies on the loopback device. The call to `update_parameters_...` fixes this new issue. Without it, after mounting again, if the same loopback device is used, we notice that the mountinfo parameters didn't change and don't trigger the update of automatic dependencies. This is not the only way this could be fixed. But I think it is actually pretty nice — I don't think we should "cache" stale information from mountinfo to save a few small allocs. In normal use the remount might never happen, or the loopback device might be different, etc. With this call, we stop the mount and destroy the volatile information, even if it is possible that it might be reused if we remount in the same place and with the same options.

Reply:
In my tests, the mountinfo data survived until the next start job. I'll try to figure out why it's not flushed.